### PR TITLE
Initial implementation for ONNX serializing

### PR DESCRIPTION
**Pull request by matwey**
Related to #27

### CODE REVIEW

File: coniferest/onnx/operator_converters/aadforest.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    attr_pairs['target_ids'].append(0)
+    attr_pairs['target_weights'].append(weight)
+
+def add_tree_to_attribute_pairs(attr_pairs, tree, tree_id, model):
+    parents = {}
+
+    for i in range(tree.node_count):
+        node_id = i
+
+        if tree.children_left[i] > i or tree.children_right[i] > i:
```
Comment (hombit) [inline coniferest/onnx/operator_converters/aadforest.py:56]: I don't understand this, could you please leave a comment why for the leaf children values are less than `i`?

File: pyproject.toml [hunk 2]
```diff
@@ -34,6 +34,10 @@ dependencies = [
   "numpy",
   "scikit-learn>=1,<2",
   "matplotlib",
+  "onnxconverter-common",
+  "skl2onnx",
+  # onnxruntime is required only for testing
```
Comment (hombit) [inline pyproject.toml:39]: Let's remove it, we don't have any other test dependencies here, like pytest. I created an issue for this
https://github.com/snad-space/coniferest/issues/107

For now, please add it to github action workflow in the same places where we install pytest

File: coniferest/onnx/shape_calculators/aadforest.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from onnxconverter_common.registration import register_shape_calculator
+from onnxconverter_common.data_types import FloatTensorType
+from onnxconverter_common.utils import check_input_and_output_types, check_input_and_output_numbers
+
+
+def calculate_aadforest_output_shapes(operator):
```
Comment (hombit) [inline coniferest/onnx/shape_calculators/aadforest.py:6]: I don't understand what is happening here. Probably it is in onnx docs somewhere, but if not, could you please briefly describe it?

File: coniferest/onnx/_parse.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+from ._container import ConiferestModelContainer
+from onnxconverter_common.topology import Topology
+from onnxconverter_common.data_types import FloatTensorType
+
+from ..aadforest import AADForest
+
+
+def _get_coniferest_operator_name(model):
+    if isinstance(model, AADForest):
```
Comment (hombit) [inline coniferest/onnx/_parse.py:9]: What if we have a subclass, do we still want this if statement match?

File: coniferest/onnx/_parse.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        return "AADForest"
+
+    raise ValueError("No proper operator name found for '%s'" % type(model))
+
+def _parse_coniferest(scope, model, inputs):
+    this_operator = scope.declare_local_operator(_get_coniferest_operator_name(model), model)
+    this_operator.inputs = inputs
+
+    # FIXME: probably another variable is required for anomality label
```
Comment (hombit) [inline coniferest/onnx/_parse.py:18]: Do we keep it for now?

File: coniferest/onnx/convert.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    return onnx_model
+
+def to_onnx(model, X=None, name=None, initial_types=None, target_opset=None,
+            targeted_onnx=onnx.__version__, custom_conversion_functions=None,
+            custom_shape_calculators=None):
+
+    if name is None:
+        name = "ONNX(%s)" % model.__class__.__name__
+
+    # FIXME: skl2onnx is required only for this helper function
```
Comment (hombit) [inline coniferest/onnx/convert.py:37]: Maybe it is ok and we can just move this import to the top? How stable skl2onnx is?

File: tests/test_onnx.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    data, _metadata = single_outlier()
+    data = data.astype(np.float32)
+    model = AADForest(n_trees=10, random_seed=0).fit(data)
+    scores_model = model.score_samples(data)
+
+    o = to_onnx(model, data)
+    onnx.checker.check_model(o)
+    scores_onnx = _evaluate_onnx(o, data)
+
+    np.testing.assert_allclose(scores_onnx, scores_model, 1e-5)
```
Comment (hombit) [inline tests/test_onnx.py:27]: Should they be absolutely equal? Or does ONNX use float32, while scikit uses float64?

**Review by hombit — CHANGES_REQUESTED** [review]
Thank you! Despite the code comments I have two other points:
1. It would be nice to have AADForest.to_onnx() method. Maybe it could save directly to the file?
2. Do we need a regression test based on a current ONNX model?

**Comment by hombit** [comment]
Please add docstring to `to_onnx` function. Maybe we also need to mention it somewhere in the documentation, but I'm not sure where exactly.

File: coniferest/onnx/_parse.py (hunk 4 above)
Comment (matwey) [reply coniferest/onnx/_parse.py:9]: This is good question. ONNX-serialized model is just a forest. It doesn't know, and doesn't need to know how it was constructed. So, subclass might work if `score_samples` are not overridden.

File: tests/test_onnx.py (hunk 7 above)
Comment (matwey) [reply tests/test_onnx.py:27]: I am not sure they can be absolutely equal.

File: coniferest/onnx/_parse.py (hunk 4 above)
Comment (hombit) [reply coniferest/onnx/_parse.py:9]: Than let's maybe restrict it to the specific class?

File: coniferest/onnx/_parse.py (hunk 5 above)
Comment (matwey) [reply coniferest/onnx/_parse.py:18]: I think so. Really, we never need anomality labels, only the scores.

File: coniferest/onnx/operator_converters/aadforest.py (hunk 1 above)
Comment (matwey) [reply coniferest/onnx/operator_converters/aadforest.py:56]: As far as I remember, I steal this code from skl2onnx :-)

**Comment by matwey** [comment]
> Thank you! Despite the code comments I have two other points:
> 
>     1. It would be nice to have AADForest.to_onnx() method. Maybe it could save directly to the file?
> 

I've tried... and we have to decide about input dtype.
I think we should allow creating models for both Double and Float.

>     2. Do we need a regression test based on a current ONNX model?

I am not sure what is the reason?

File: pyproject.toml (hunk 2 above)
Comment (hombit) [reply pyproject.toml]: I made it in https://github.com/snad-space/coniferest/pull/108, so now we have a special section of `pyproject.toml` to add dev dependencies to

File: coniferest/onnx/operator_converters/aadforest.py (hunk 1 above)
Comment (hombit) [reply coniferest/onnx/operator_converters/aadforest.py:56]: @matwey do we have the same tree implementation as there? I'm ok about it we now what we are doing...

**Comment by hombit** [comment]
I see very little reason for Double models, especially for trees. The only thing which could change because of rounding errors is probably scoring, but even there we do not have millions of trees... But if it is simple to do, let's do it!

**Pull request merged**